File: Python/BinarySearch.py
class BinarySearch:
	def __init__(self):
		pass

	def binary_search_recursive(self, a_list, start, end, value):
		"""
		This function performs a recursive binary search on the a_list array for value
		Returns the index of the value in array if found, otherwise return -1
		"""
		mid = (start + end) / 2
		if value == a_list[mid]:
			return mid
		if start > end:
			return -1
		if value < a_list[mid]:
			return binary_search_recursive(a_list, start, mid-1, value)
		else:
			return binary_search_recursive(a_list, mid+1, end, value)

	def binary_search_iterative(self, a_list, start, end, value):
		"""
		This function performs an iterative binary search on the a_list array for value
		Returns the index of the value in array if found, otherwise return -1
		"""
		mid = (start + end) / 2
		while(start <= end):
			if value == a_list[mid]:
				return mid
			if value < a_list[mid]:
				end = mid - 1
			else:
				start = mid + 1
		return -1

if __name__ == "__main__":
	pass

File: JAVA/BinarySearch.java
import java.util.*;

public class BinarySearch {
	public int binarySearchRecursive(int[] searchArray, int start, int end, int key) {
		// This method searches a binary tree for a value recursively
		// Returns the index of the key if key found in array or return -1

		int mid = (start + end) / 2;
		if (key == searchArray[mid])
			return mid;
		if (start > end)
			return -1;
		if (key < searchArray[mid]) {
			return binarySearchRecursive(searchArray, start, mid-1, key);
		} else {
			return binarySearchRecursive(searchArray, mid+1, end, key);
		}
	}

	public int binarySearchIterative(int[] searchArray, int start, int end, int key) {
		// This method searches a binary tree for a value iteratively		
		// Returns the index of the key if found in array, otherwise return -1

		int mid = (start + end) / 2;
		while (start <= end) {
			if (key == searchArray[mid])
				return mid;
			if (key < searchArray[mid]) { // key in first part
				end = mid - 1;
			} else { // key in second part
				start = mid + 1;
			}
		}
		return -1; // key not found
	}
}